src/lib/zs_matcher.rs

```diff
@@ -37,27 +37,104 @@
 
 #![warn(missing_docs)]
 
+use std::collections::HashMap;
 use std::fmt::Debug;
 use std::f64;
+use std::hash::Hash;
 
 use ast::{Arena, FromNodeId, NodeId, ToNodeId};
-use matchers::{MappingStore, MatchTrees};
+use matchers::{has_same_type, MappingStore, MappingType, MatchTrees};
 use qgram::trigram_distance;
 
-/// The `ZhangShasha` matcher does not require any configuration.
+/// Z-S Configuration includes tree and forest distance matrices.
+///
+/// The `src` and `dst` trees are indexed in left-right postorder numbering,
+/// as described in Section 3.1 of the paper.
 #[derive(Debug, Clone, PartialEq)]
-pub struct ZhangShashaConfig {}
-
-impl Default for ZhangShashaConfig {
-    fn default() -> ZhangShashaConfig {
-        ZhangShashaConfig {}
-    }
+pub struct ZhangShashaConfig {
+    src: ZSTree<FromNodeId>,
+    dst: ZSTree<ToNodeId>,
+    tree_dist: Vec<Vec<f64>>,
+    forest_dist: Vec<Vec<f64>>
 }
 
 impl ZhangShashaConfig {
     /// Create a new configuration object.
-    pub fn new() -> ZhangShashaConfig {
-        Default::default()
+    pub fn new<T: Clone + PartialEq + ToString>(src: &Arena<T, FromNodeId>,
+                                                dst: &Arena<T, ToNodeId>)
+                                                -> ZhangShashaConfig {
+        ZhangShashaConfig { src: ZSTree::new(src),
+                            dst: ZSTree::new(dst),
+                            tree_dist: vec![vec![]],
+                            forest_dist: vec![vec![]] }
+    }
+
+    /// Compute the tree distance matrix. See Section 3.2 of the paper.
+    fn compute_tree_distance<T: Clone + Debug + Eq + ToString + 'static>(&mut self,
+                                                                         store: &MappingStore<T>)
+    {
+        self.tree_dist =
```

Comment:
Couldn't these two assignments be put in the `for` loop immediately afterwards? That would be easier to read and, probably, faster.